src/Microsoft.IdentityModel.Tokens/Cache/MaximumSizeCache.cs

```diff
@@ -0,0 +1,272 @@
+        /// </summary>
+        /// <param name="key">The key of the value to get or set.</param>
+        /// For get, if the specified key is not found, a <exception cref="KeyNotFoundException"> will be thrown.</exception>
+        /// For set, if the key is null throw an <exception cref="ArgumentNullException"> will be thrown.</exception>
+        public CacheItem<TKey, TValue> this[TKey key]
+        {
+            get => _map[key];
+            set => _map[key] = value;
+        }
+
+        /// <inheritdoc/>
+        public virtual bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        /// <inheritdoc/>
+        public virtual void SetValue(TKey key, TValue value)
+        {
+            SetValue(key, value, DateTime.MaxValue);
+        }
+
+        /// <inheritdoc/>
+        public virtual bool SetValue(TKey key, TValue value, DateTime expirationTime)
+        {
+            ValidateValues(key, value, expirationTime);
+            if (NeedsCompaction && Interlocked.CompareExchange(ref _compactionState, CompactionStateInProgress, CompactionStateNotRunning) == CompactionStateNotRunning)
+                ThreadPool.QueueUserWorkItem(CompactCache);
+
+            _map[key] = new CacheItem<TKey, TValue>(key, value, expirationTime);
+
+            return true;
+        }
+
+        /// <inheritdoc/>
+        public virtual bool TryGetValue(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            value = default;
+            var found = TryGetValueInternal(key, out CacheItem<TKey, TValue> val);
+            if (found)
+                value = val.Value;
+
+            return found;
+        }
+
+        /// <inheritdoc/>
+        public virtual bool TryRemove(TKey key, out TValue value)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            value = default;
+
+            var found = TryRemoveInternal(key, out CacheItem<TKey, TValue> val);
+            if (found)
+                value = val.Value;
+
+            return found;
+        }
+
+        #endregion
+
+
+        #region protected
+
+        /// <summary>
+        /// The internal version of TryGetValue that gets the LRUCacheItem from the hash.
+        /// </summary>
+        /// <param name="key">The key of the value to get.</param>
+        /// <param name="value">The object that has the specified key, or the default value of the type if the not found.</param>
+        /// <returns>true if the key was found, otherwise, false.</returns>
+        protected virtual bool TryGetValueInternal(TKey key, out CacheItem<TKey, TValue> value)
+        {
+            value = default;
+
+            var found = _map.TryGetValue(key, out CacheItem<TKey, TValue> val);
```

Comment:
could you write this as:
```cs
return _map.TryGetValue(key, out CacheItem<TKey, TValue> value);
```

Reply:
The TryGetInternal() and TryRemoveInternal() return a CacheItem, not just the value of the CacheItem. A CacheItem has the ExpirationTime that the EventBasedLRUCache needs.

Maybe the methods should be renamed to TryGetCacheItem() and TryRemoveCacheItem()?